pilot/pkg/security/authn/v1beta1/policy_applier.go

```diff
@@ -281,3 +307,68 @@ func convertToEnvoyJwtConfig(jwtRules []*v1beta1.JWTRule) *envoy_jwt.JwtAuthenti
 		Providers: providers,
 	}
 }
+
+// GetMutualTLSMode returns the MutualTLSMode enum corresponding to the given peer authentication policy.
+// If the input authentication is nil, returns MTLSPermissive.
+func GetMutualTLSMode(peer *v1beta1.PeerAuthentication) model.MutualTLSMode {
+	if peer == nil || peer.Mtls == nil {
+		return model.MTLSPermissive
+	}
+	switch peer.Mtls.Mode {
+	case v1beta1.PeerAuthentication_MutualTLS_DISABLE:
+		return model.MTLSDisable
+	case v1beta1.PeerAuthentication_MutualTLS_PERMISSIVE:
+		return model.MTLSPermissive
+	case v1beta1.PeerAuthentication_MutualTLS_STRICT:
+		return model.MTLSStrict
+	default:
+		// Should never reach here
+		log.Errorf("Unknown peer authentication %#v", peer)
+		return model.MTLSUnknown
+	}
+}
+
+// GetMostSpecificScopeConfig returns the config with the most specific scope, i.e config with
+// non-empty selector should be preferred over those without, one in non-root namespace should be
+// preferred over those in root namespace. Where there are more than one config in the same
+// scope, pick the one with stronger security (STRICT > PERMISSIVE > DISABLE). In a tie, the first
+// in the list will be used (typically, the list is passed in sorted order).
+// The input configs is assumed to be a short list of configs that should be considerred for an
+// workload.
+func GetMostSpecificScopeConfig(rootNamespace string, configs []*model.Config) *model.Config {
+	// 0 - workload-specific config, define in the same namespace as workload (non-root namespace).
+	// 1 - workload-specific config, define in the root namespace.
+	// 2 - namespace level config (i.e config without selector, in non-root namespace)
+	// 3 - mesh level config (i.e config without selector, in root-namespace)
+	configByScope := make([]*model.Config, 4)
+
+	for _, cfg := range configs {
+		spec := cfg.Spec.(*v1beta1.PeerAuthentication)
+		level := 0
+		if cfg.Namespace == rootNamespace {
+			level = 1
+		}
+		if spec.Selector == nil || len(spec.Selector.MatchLabels) == 0 {
+			level += 2
+		}
+		if isStrictlyStronger(spec, configByScope[level]) {
+			configByScope[level] = cfg
```

Comment:
This logic is memory intensive, allocating an array and destroying it for each query for each proxy in the system. It will also undergo a lot of changes when you introduce workload specific sidecars. To me, a better way is to reuse the logic that builds https://github.com/istio/istio/blob/master/pilot/pkg/model/push_context.go#L95 `sidecarsByNamespace` . And then as commented above, you simply have to query push context for the namespace's peer authN policy. Otherwise, this will become a performance bottleneck once again.

Reply:
We already use that. This is a second round to pick from the short list (the input list is a list of config that match the workload) based on our precedent order rules.

Reply:
You dont need the list of configs - that is my point. Take a look at the code in push context. It simply scans through a hash map and picks the appropriate config without allocating more memory. If there is no namespace level config, it automatically falls back to the global config which *always* exists even if the user did not define a global setting.

Reply:
There are few things that the approach with sidecar doesn't work here:
- We want to take into account the 'strength' of mTLS mode to break the tie.
- We need to 'inherit' from parent for UNSET mode (thanks to your other comment :) )

I've updated this PR to handle UNSET, but I'm still working on simplify it a bit more, using the assumption that there is at most one namespace & mesh level policy (we already enforce that from validation), and also ignore policies in root namespace but have selector (as you suggested)

Reply:
This is the pointer of array, `[]*model.Config`, not the `[]model.Config`, will that still be mem intensive concern?

Reply:
I re-implement this function to take into account UNSET (which should be inherit from parent), and port-level settings (which need to be aggregated to the final policy, including resolve for UNSET). It seems more complicated now, unfortunately, though I think it has the correct behavior.

PTAL